### ISSUE TITLE
Ratio

### ISSUE THREAD

**Issue by factor01211984** [UncleMeat/gazelle#4]
Hello friends it is me again Manish. In my testing I am unable to find where to set ratio limits for required ratio. Is it possible to do this?

**Comment by ValliusDax** [comment]
gazelle/sections/schedule/index.php

criteria array

**Comment by factor01211984** [comment]
Sorry brother, I attempted to edit this in promotions array but there is no visible effect. The only change I make is to change user classes to Hindi with english characters of course but it does not seem to change required ratio in user profile. Would changing the name of user classes negate effect?

**Comment by ValliusDax** [comment]
He careful how you edit that array. If you change those "class names" then you must also change them in the config file (classes/config.php) and make sure the numbers assigned to them match the PermissionID of those classes in the DB.

Secondly, there is more than one side to the required ratio. There is the enforcing of the ratio, displaying the ratio, and promotions/demotions due to ratio.

All of this is handled by the scheduler (sections/schedule/index.php), I suggest you do some reading.

**Comment by factor01211984** [comment]
Thank you brother, I will edit the class names neatly and try again. After some time then I will report to you the result.

**Comment by factor01211984** [comment]
Sorry one more question, are class names case sensitive? I observe that class names in config.php and in schedule/index.php are all in english uppercase, also class names which are more than two words are spaced with _ is this nessecary for proper function?

**Comment by ValliusDax** [comment]
I hate to do this, but php is very well documented, so RTFM.

http://php.net/manual/en/function.define.php

**Comment by factor01211984** [comment]
Excellent! I appreciate this. So capitals are case sensitive. But this also meas user classes in database and on site must be defined in capital letters correct?

**Comment by ValliusDax** [comment]
No, the user classes as defined in the config file are really just names which relate to numbers. Those numbers must match the PermissionID like I said previously.

Consider this as an example permissions table and class definition:

``` SQL
mysql> SELECT ID, Name FROM permissions;
+----+------------+
| ID | Name       |
+----+------------+
|  1 | Admin      |
|  2 | User       |
|  3 | Good User  |
|  4 | Uploader   |
+----+------------+
```

``` php
// Name of class    Class ID (NOT level)
define('ADMIN',     '1'); 
define('USER',      '2'); 
define('GOOD_USER', '3'); 
define('UPLOADER',  '4');
```

**Comment by factor01211984** [comment]
Excellent brother! I will go over the provided reading and make my attempts. I will report to you my results. Thank you very much!

**Comment by factor01211984** [comment]
Still here brother. Nothing yet to report...

**Comment by factor01211984** [comment]
Hello again, it is me Manish. I am testing for two days and I think I find my problem. For lowest caste I chyange from apprentice to Aspr̥sya. 

Site will change castes when appropriate statistics are entered into database but I observe required ratio in profile of all castes is 0.00.

Also lowest caste will not promote from Aspr̥sya to Shudra. No matter how I manipulate statistics. 

Finally I also observe that site stats will not update unless I reboot machine. Then is only update once. Any thoughts brother. Your advise is appreciated.

**Comment by ValliusDax** [comment]
Have you configured your cronjobs?

Add the following to your crontab

```
/usr/bin/php <path to gazelle>/schedule.php <schedule key you set in config.php>
```

For instance

```
/usr/bin/php /var/www/gazelle/schedule.php abcdefghijklmn12345
```

You can of course also trigger this manually by running the command in a console or by selecting the "schedule" tool from the tools page (warning, it takes a long time to run and may timeout in a browser).

**Comment by factor01211984** [comment]
I will check this. Please to note. I am sure is not a problem with your code. Code is excellent. I have made an error in something. Thank you for being so kind to help me to find my mistake.

**Comment by factor01211984** [comment]
Okay brother I think we have found problem. In cron tab I have:

0,15,30,45 \* \* \* \* php /var/www/schedule.php abcdefghijk1234567>> /home/USER/logs/schedule.log

This is obvious syntax deviation from your suggestion. So I should change what I have above to

/usr/bin/php <path to gazelle>/schedule.php <schedule key you set in config.php>

Or simply add your suggestion?

**Comment by ValliusDax** [comment]
No, the stuff at the start is needed, I should have included that previously.

Check the path to schedule.php is right and that the password you have is right. If they both check out then check the schedule.log to see if there are any errors. If it's all clear then the issue is with your code and it is not correctly updating the DB.

**Comment by factor01211984** [comment]
Okay, while tinkering I am now face with new problem:

MySQL: Invalid Query: SELECT ID, Name, Description, Level, Color, LOWER(REPLACE(Name,' ','')) AS ShortName, IsUserClass FROM permissions ORDER BY IsUserClass, Level SQL error: 1054 (Unknown column 'Description' in 'field list')

Array ( [0] => Array ( [0] => SELECT ID, Name, Description, Level, Color, LOWER(REPLACE(Name,' ','')) AS ShortName, IsUserClass FROM permissions ORDER BY IsUserClass, Level [1] => 1.9431114196777 ) ) 

Do you rcognise this error?

**Comment by factor01211984** [comment]
Nevermind I see it now...

**Comment by factor01211984** [comment]
Hello brother, again is Manish. Unfortunate to report still no success. I am replace all of code with unmodified original and let run for ome hours. schedule.log is report 1 error

<pre>MySQL: Invalid Query: DELETE FROM ip_bans WHERE EndTime!=&#39;0000-00-00 00:00:00&#39; AND EndTime&lt;&#39;2015-07-16 23:00:06&#39; SQL error: 1054 (Unknown column &#39;EndTime&#39; in &#39;where clause&#39;)</pre>

 I hope maybe you can recognise this errors.

**Comment by factor01211984** [comment]
Hello again, is me Manish. I am sorry to be so much trouble but I still cannot solve this problem. I have recompiled the code and made the suggested adjustments but the issue remains the same.

The required ratio field in user profile as I show here:

http://i.imgur.com/KeTf8NC.jpg

Remains unchanged, and user is still not promoted. The errors in schedule log are as follows:

<pre>MySQL: Invalid Query: INSERT INTO temp_copy (uid, downloaded, remaining, uploaded, upspeed, downspeed, timespent, peer_id, ip, fid, mtime)
                    SELECT x.uid, x.downloaded, x.remaining, x.uploaded, x.upspeed, x.downspeed, x.timespent, x.peer_id, x.ip, x.fid, x.mtime
                      FROM xbt_peers_history AS x
                 LEFT JOIN users_watch_list AS uw ON uw.UserID=x.uid
                 LEFT JOIN torrents_watch_list AS tw ON tw.TorrentID=x.fid
                     WHERE uw.UserID IS NOT NULL
                        OR tw.TorrentID IS NOT NULL
                        OR x.upspeed &gt;= &#39;&#39;
                        OR x.mtime&gt;&#39;1437385512&#39; 
SQL error: 1146 (Table &#39;gazelle.temp_copy&#39; doesn&#39;t exist)</pre>Array
 If allowed to run error is changed to this:
<pre>MySQL: Invalid Query: CREATE TABLE `temp_copy` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `uid` int(11) NOT NULL,
  `downloaded` bigint(20) NOT NULL,
  `remaining` bigint(20) NOT NULL,
  `uploaded` bigint(20) NOT NULL,
  `upspeed` bigint(20) NOT NULL,
  `downspeed` bigint(20) NOT NULL,
  `timespent` bigint(20) NOT NULL,
  `peer_id` binary(20) NOT NULL DEFAULT &#39;^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@&#39;,
  `ip` varchar(15) NOT NULL DEFAULT &#39;&#39;,
  `fid` int(11) NOT NULL,
  `mtime` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  KEY `uid` (`uid`),
  KEY `fid` (`fid`),
  KEY `upspeed` (`upspeed`),
  KEY `mtime` (`mtime`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 
SQL error: 1050 (Table &#39;temp_copy&#39; already exists)</pre>

I assume table is being created but I cannot find in database. But original problem remains. How to properly set user required ratio. Also why is user not being promoted even if tinker with stats. I check whatcd version and user classes are obviously different from this version so I am sure this is possible. I am very hopeful someone will have a solution for this issue. Thank you.

**Comment by ValliusDax** [comment]
You need to learn SQL and PHP, there is a limit to the support we can give and you have reached this limit. Please go and read a book about the technologies you are struggling to understand.

**Comment by factor01211984** [comment]
I am sorry to be so much trouble. It would seem that there is a lot to learn to make such a simple adjustment. If I may ask just one more question, perhaps for a suggestion of a starting point?

**Comment by ValliusDax** [comment]
There are a lot of books for PHP and SQL, both are very well documented mature languages so just about any book from a reputable publisher would do.

If you don't want to spend money then there are also a lot of online tutorials and such. Try using google.

**Comment by factor01211984** [comment]
I am sorry, I mean starting point to address this issue. It would seem a lot to learn for a simple change. But I do not wish to trouble you any further. Thank you very much for all of your help.

**Comment by ValliusDax** [comment]
The SQL creating those errors is at the bottom of the scheduler/index.php file, it should not cause those and without knowing how you have altered the code I could not tell you how you have broken it.

Again, gaining general knowledge about PHP and SQL is the best way to address this, such knowledge will help you to debug issues such as this without needing to resort to asking for help.

**Comment by factor01211984** [comment]
Thank you very much for your kindness. I will investigate this. You have provided a great service to the world and I appreciate you. Thank you again.
